### ISSUE TITLE
CMake questions and suggestions

### ISSUE THREAD

**Issue by Lwmte** [opentomb/OpenTomb#153]
What's the difference between upper/lower case in function names? E.g., sometimes, in CMakeLists.txt, I see ` FIND_PACKAGE(blah) ` and sometimes I see ` find_package(blah) ` .

Also, forgive me for my stupidness, but why SDL is added via complicated workaround, not just like with any other libraries (excluding OpenAL, which we discussed in #115)?

Next, can we add some kind of flag to make a choice - link libraries dynamically or statically?

**Comment by vvs-** [comment]
In documentation the commands are always lower case and the variables are upper case. I don't know if cmake is case sensitive.

I believe that you can add any variable to a project and use it to distinguish between different setups.

**Comment by vvs-** [comment]
From the manual: `Command names are case-insensitive.`

**Comment by stohrendorf** [comment]
... and variables are *not*. And in `if` expressions, variable names are automatically evaluated, so in `IF(LUA_FOUND)`, the variable `LUA_FOUND` gets evaluated. When there is a variable `FOO` that contains `"LUA_FOUND"`, an `IF(${FOO})` expands to `IF(LUA_FOUND)`. CMake has its pitfalls, but it's still better than using IDE-bound (and platform-bound) project files...

**Comment by stohrendorf** [comment]
BTW, always link libs *shared*. On windows systems, you have to distribute them in a specific version anyway, as there is no common libs folder (except system32, but we don't really want to use that). On *nix compatible systems, there are always versioned symlinks to the respective .so's, so there shouldn't be a problem either. Yet if there are external libraries specifically built *inside* the source (like OpenAL here for Mac OS X), it should *always* be linked static to avoid that library hell.

**Comment by Lwmte** [comment]
Simply said, it's better to deploy 10 DLLs with engine than integrate them into executable?

**Comment by vvs-** [comment]
Well, on Linux you don't have to. That means that there are different requirements for different platforms.

**Comment by Lwmte** [comment]
Maybe it's OK for the user, I don't know... Recently I saw a demo from pouet.net with LOTS of dlls attached, it just looked so unusual, as I used to see only one executable in folder! :smile: I just think there should be a choice between dynamic and static for Windows machines.

**Comment by Lwmte** [comment]
I have another question - is it possible to add all script files to CodeBlocks/Eclipse/VS project generated by CMake, but ignore them for generating makefiles themselves?

**Comment by Lwmte** [comment]
Made some CMakeLists.txt refactoring - let's hope I didn't break anything this time! :smile:

**Comment by Gh0stblade** [comment]
Windows CMake build tutorial pls (Cmake newbie)

**Comment by Lwmte** [comment]
Something I found on quick googling: https://secure.mash-project.eu/wiki/index.php/CMake:_Getting_Started (although it's for specific project, but general info is applicable to OpenTomb as well).

However, I didn't read *any* manuals at all to get used to it! It's very simple. You just install CMake, launch CMake-gui, point it to source code folder and target folder (where it generates all needed stuff), push "Configure" and then "Generate" button (you maybe have to press "Configure" button twice, as it will yell at you at first). CMake then offers you different generators at your choice (I used CodeBlocks project generator). Then it generates project file and makefile, you just load this project and then can compile engine almost without any extra setup (I just needed to change build type in CodeBlocks from "Comments only" to "GUI application").
